lib/assert.js

```diff
@@ -608,6 +612,142 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+/**
+ * Compares two objects or values recursively to check if they are equal.
+ * @param {any} actual - The actual value to compare.
+ * @param {any} expected - The expected value to compare.
+ * @param {boolean} [loose=false] - Whether to use loose comparison (==) or strict comparison (===). Defaults to false.
+ * @param {Set} [comparedObjects=new Set()] - Set to track compared objects for handling circular references.
+ * @returns {boolean} - Returns `true` if the actual value matches the expected value, otherwise `false`.
+ * @example
+ * // Loose comparison (default)
+ * compareBranch({a: 1, b: 2}, {a: 1, b: '2'}); // true
+ *
+ * // Strict comparison
+ * compareBranch({a: 1, b: 2}, {a: 1, b: 2}, true); // true
+ */
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new Set(),
+) {
+  function isPlainObject(obj) {
+    if (typeof obj !== 'object' || obj === null) return false;
+    const proto = ObjectGetPrototypeOf(obj);
+    return proto === Object || proto === null;
```

Comment:
Actually my proposal still fails to account for different realms:

```console
$ ./node -e 'assert.matchObjectStrict(vm.runInNewContext("({})"), {})'
node:assert:131
  throw new AssertionError(obj);
  ^

AssertionError [ERR_ASSERTION]: {} matchObjectStrict {}
    at [eval]:1:8
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:118:14
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:101:62)
    at evalScript (node:internal/process/execution:136:3)
    at node:internal/main/eval_string:49:3 {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: {},
  expected: {},
  operator: 'matchObjectStrict'
}

Node.js v23.0.0-pre
```
```suggestion
    const proto = ObjectGetPrototypeOf(obj);
    return proto === ObjectPrototype || proto === null || ObjectPrototypeToString(obj) === '[object Object]';
```

Can we add a test for objects coming from a different realm?